### ISSUE TITLE
Solve Richards Equation using H(div) FE space

### ISSUE THREAD

**Issue by rezgarshakeri** [mfem/mfem#2660]
I'd like to add an example to mfem to solve Richards equation.
The Richards equation can be found [here](https://www.pflotran.org/documentation/theory_guide/mode_richards.html).
cc: @jedbrown

**Comment by pazner** [comment]
Can this be done using a modification of the [Darcy problem example 5](https://mfem.org/examples/#example-5-darcy-problem), which uses H(div) RT elements?

**Comment by rezgarshakeri** [comment]
That is our plan. First I need to add a time derivative term (dp/dt) to the second equation of example5, then implement the complete form of Richards equation.

**Comment by v-dobrev** [comment]
Hi @rezgarshakeri,

I invited you to the MFEM/contributors team. You should be able to create your branch directly in the MFEM/mfem repo. Let me know if you encounter any issues with permissions or anything else.

**Comment by rezgarshakeri** [comment]
Hi @v-dobrev ,
Thanks for your invitation. I have created my branch ('rezgar/RichardsEquation').

**Comment by rezgarshakeri** [comment]
@v-dobrev is there any time dependent MMS in mfem?

**Comment by rezgarshakeri** [comment]
I am trying to solve the problem using backward Euler (as Jed suggested to keep the symmetry of the problem), now I have   

[ M  B^T ] [u(t_n+1)] = [f(t_n+1)] + [0     0][u(t_n)]
[ B  C/dt] [p(t_n+1)] = [g(t_n+1)] + [0  C/dt][p(t_n)]
 where:
M = \int_\Omega k u_h \cdot v_h d\Omega   u_h, v_h \in R_h
B   = -\int_\Omega \div u_h q_h d\Omega   u_h \in R_h, q_h \in W_h
C   = -\int_\Omega p_h q_h d\Omega        p_h \in W_h, q_h \in W_h

**Comment by v-dobrev** [comment]
Hi @rezgarshakeri,

I'm not sure if the above comments were meant just as a status update, or not. If you need help to proceed, can you elaborate what part of the algorithm is the problem?

**Comment by rezgarshakeri** [comment]
Hi @v-dobrev ,
Thanks for your comment. That comment is just and an update. I am trying to implement backward Euler as I wrote it above. One question I have is, can we compute `C*p(t_n)/dt` to avoid creating those block zero matrices shown above? 
Sorry for not updating so long, I am implementing H(div) space using BDM element in libCEED and later I'll compare it with RT element in mfem.

**Comment by v-dobrev** [comment]
> One question I have is, can we compute C*p(t_n)/dt to avoid creating those block zero matrices shown above?

Since you assemble `C` (or `C/dt`), you can compute the r.h.s. contribution `C*p(t_n)/dt` for the equations at `t=t_{n+1}` by sparse matrix-vector multiplication since you have the vector `p(t_n)` from the previous time step.

**Comment by v-dobrev** [comment]
> Sorry for not updating so long, I am implementing H(div) space using BDM element in libCEED and later I'll compare it with RT element in mfem.

No problem at all, there's no rush. It will be interesting to see what you find in your comparison. We can certainly add BDM to MFEM too, if there is interest.

**Comment by rezgarshakeri** [comment]
Hi @v-dobrev , 
I did matrix-vector multiplication as you said, thanks.
Now I need to implement the delta function as point source to the problem. In fact, I need to solve
k*u + grad p = grad (\rho g z)
-div u = delta (r-ro) + dp/dt

Is there a way to apply point-force in mfem?

(I also implemented linear BDM for quad and hex elements in libCEED, and solved the Darcy problem (mfem/ex05) and got the convergence rate, I need to add some visualization feature to see the output in paraview)
